airflow-core/src/airflow/jobs/scheduler_job_runner.py

```diff
@@ -2108,18 +2136,43 @@ def _emit_pool_metrics(self, session: Session = NEW_SESSION) -> None:
         with DebugTrace.start_span(span_name="emit_pool_metrics", component="SchedulerJobRunner") as span:
             pools = Pool.slots_stats(session=session)
             for pool_name, slot_stats in pools.items():
-                Stats.gauge(f"pool.open_slots.{pool_name}", slot_stats["open"])
-                Stats.gauge(f"pool.queued_slots.{pool_name}", slot_stats["queued"])
-                Stats.gauge(f"pool.running_slots.{pool_name}", slot_stats["running"])
-                Stats.gauge(f"pool.deferred_slots.{pool_name}", slot_stats["deferred"])
-                Stats.gauge(f"pool.scheduled_slots.{pool_name}", slot_stats["scheduled"])
-
-                # Same metrics with tagging
-                Stats.gauge("pool.open_slots", slot_stats["open"], tags={"pool_name": pool_name})
-                Stats.gauge("pool.queued_slots", slot_stats["queued"], tags={"pool_name": pool_name})
-                Stats.gauge("pool.running_slots", slot_stats["running"], tags={"pool_name": pool_name})
-                Stats.gauge("pool.deferred_slots", slot_stats["deferred"], tags={"pool_name": pool_name})
-                Stats.gauge("pool.scheduled_slots", slot_stats["scheduled"], tags={"pool_name": pool_name})
+                # If enabled on the config, publish metrics twice,
+                # once with backward compatible name, and then with tags.
+                DualStatsManager.gauge(
+                    f"pool.open_slots.{pool_name}",
+                    "pool.open_slots",
+                    slot_stats["open"],
+                    tags={},
+                    extra_tags={"pool_name": pool_name},
+                )
```

Comment:
I didn't check all cases in this PR, but I wonder if we could do this as 
```suggestion
                DualStatsManager.gauge(
                    "pool.open_slots",
                    slot_stats["open"],
                    tags={},
                    extra_tags={"pool_name": pool_name},
                )
```

And then we can build the old name via something like this:

```python
metric_name = metric + ".".join(extra_tags.values())
```

wdyt?

Reply:
We could iterate over the tags and for each tag, add a dot and the value in the name. That would standardize it, but the legacy names don't always follow the same format.

For example, in the `dagrun` the legacy name contains just the `dag_id` but the tags have extra info.

Here are some examples,

https://github.com/apache/airflow/blob/e4c01c3b330c68feff7cd7dee0983858e6d6c2d0/airflow-core/src/airflow/models/dagrun.py#L1723-L1728

https://github.com/apache/airflow/blob/e4c01c3b330c68feff7cd7dee0983858e6d6c2d0/airflow-core/src/airflow/models/dagrun.py#L1153-L1158

and these are the tags

https://github.com/apache/airflow/blob/e4c01c3b330c68feff7cd7dee0983858e6d6c2d0/airflow-core/src/airflow/models/dagrun.py#L426-L427

Reply:
Maybe then a `legacy_name` that defaults to None and is generated in that case? That would make most cases simpler?

Reply:
When it defaults to `None`, we will append all available tags to the legacy_name? 
 
My preference would be to remove the legacy names entirely but I don't know how many people are using the names in their workflows to gather stats. That's why I tried to keep the exports the same and tried to make the existing tests work with the new code.

Your suggestion will make the code simpler but it might not be that easy to maintain exactly the same metrics as before.

Reply:
When I considered this problem in the past I had a half-baked idea of creating a "metrics registry" (a yaml file?) which would serve the purpose of being used to build the metrics documentation page so it was always up to date, and also providing a regex pattern for the old name conversion.  I don't know if you want to attempt that approach or not, but it's (half of) a thought.  Something like:

- metrics
  - metric_name_1
    - description
    - type
    - old_name_pattern
  - metric_name_2  
  

Then the table at https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/logging-monitoring/metrics.html#metric-descriptions could be built from that and we could use a static check to make sure that yaml (and therefor the docs) are always up to date.

Reply:
> also providing a regex pattern for the old name conversion.

@ferruzzi Do you mean forcing all legacy metric names to follow the same convention?

Right now, it goes like this

* name: `<metric_name>`
* legacy: `<metric_name>.<variables>`

In place of `<variables>` we could have any number of values. It could be one or more, all connected with dots. E.g. `<metric_name>.dag_id` or `<metric_name>.dag_id.task_id`.

Reply:
No, I was thinking we could maybe have a default regex pattern, but not all legacy names follow the same naming convention so the legacy_name_pattern would let us define the regex pattern for each, or only for the ones that don't follow the default.

Reply:
> or only for the ones that don't follow the default.

I think you mean that we can't define a pattern for the ones that don't follow the default naming convention.

Or do you mean define a pattern so that they don't deviate much and make sure that the pattern is followed?

Reply:
LOL  Sory, I'm not explaining my idea very well I guess.  Alright, back this up a bit.  Forget everything about a possible default, that was just adding confusion.

The registry entries for `ti.finish.<dag_id>.<task_id>.<state>` and `task_restored_to_dag.<dag_id>` might look like this:

```
metrics:
  - name: "ti.finish"
    description: "Number of completed task in a given Dag. Similar to <job_name>_end but for task"
    type: "counter"
    legacy_name: "ti.finish.{dag_id}.{task_id}.{state}"

  - name: "task_restored_to_dag"
    description: "Number of tasks restored for a given Dag (i.e. task instance which was previously in REMOVED state in the DB is added to Dag file). Metric with dag_id and run_type tagging."
    type: "counter"
    legacy_name: "task_restored_to_dag.{dag_id}"

```

then in the code before emitting the metric, we populate the legacy_name template using the metric's `tags`.  

As an added bonus, as part of the build-docs script we can walk through this registry and generate the table [in the docs](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/logging-monitoring/metrics.html) so that doc page is always up to date.

Reply:
We could also use the registry to address the initial comment of this thread and simplify the usage of the `DualStatsManager` calls. Each method will only accept the metric name and we will look it up in the registry and get the legacy name from there. @ashb Wdyt?

Reply:
@ferruzzi I made the changes. The metric tables are generated from a common yaml file. I replaced the syntax `<variable>` with `{variable}` because we can do something like this

```python
legacy_name = "ti.finish.{dag_id}.{task_id}.{state}"
formatted_legacy_name = legacy_name.format(dag_id="my_dag", task_id="my_task1", state="success")
```

I'm not sure if it's worth it. We could end up writing way more code and complicating things with no significant benefit. I'll do a POC first and see how it looks.

Reply:
I updated the patch. All legacy metrics are accessed through the yaml file. I also added some validations to make sure that the legacy names always exist on the yaml and that we have declared the variables needed for the name.

I built the yaml using the existing tables from the `metrics.rst`. The failing tests on the CI have to do with the new validations. It seems that some metrics didn't exist in the old docs. I'm going to add them to the yaml and also fix any other failures.

Reply:
Sorry I missed you previous comment from last week.  The way you mentioned you were going to try is exactly my thought as well.

> The failing tests on the CI have to do with the new validations. It seems that some metrics didn't exist in the old docs

That's exactly what I was hoping to catch and fix with the registry idea.  There needs to be a check the other way as well; just because it is in the docs doesn't mean it still exists in the code... but we can do that part later if you want.  That might just be an initial manual cleanup plus a pre-commit hook that checks the commit for changes, but that may be a side project on its own.

Reply:
Right now I added a validation just for when reading the legacy metric from the yaml.

A pre-commit check or a cross-check validation everytime we try to add a metric to the metric_map would be good to have. But I think that it should happen in another PR because it would widen the scope of this one by a lot.

Just to get the tests working, I'm adding metrics that have been missing from the docs. The issue is that I'm not entirely sure about the description. I'm reading the code but still what I'm writing is a guess. So the validation PR, should start as a draft with a list of newly documented metrics so that people can add comments and explain what each does.

Reply:
> I think that it should happen in another PR because it would widen the scope of this one by a lot.

Yeah, I'm happy with that.  Sorry to blow the scope out on this one already.  I'm working through the changes but have a lot of meetings today so i may not get through it all today.

Thanks for sticking with this one, I know it's been a pretty long haul

Reply:
Don't worry about it. I'm also not done with the CI errors. Thanks!